lib/Graph/Grad.cpp

```diff
@@ -190,6 +189,43 @@ Function *glow::differentiate(Function *F, const TrainingConfig &conf,
       // calculations.
       continue;
 
+    if (N->getKind() == Kind::BatchNormalizationNodeKind) {
+      auto *BN = cast<BatchNormalizationNode>(N);
+      auto in = BN->getInput();
+      auto mean = BN->getMean();
+      auto var = BN->getVar();
+      auto channelIdx = BN->getChannelIdx();
+      auto momentum = BN->getMomentum();
+
+      // Update the mean and variance via the MeanVarNormalizationNode.
+      auto *MVN = new MeanVarNormalizationNode("mean_var_normalization", in,
+                                               mean, var, channelIdx, momentum);
+      toAppend.push_back(MVN);
+
+      // Save the newly calculated mean and variance to the mean and variance
+      // variables. These will be used during the next iteration of training.
+      G->createSave("saveMean", MVN->getNewMean(),
+                    llvm::cast<Variable>(mean.getNode()));
+      G->createSave("saveVar", MVN->getNewVar(),
+                    llvm::cast<Variable>(var.getNode()));
+
+      // Replace the BN's mean and variance with the new mean and variance
+      // calculated from MVN.
+      for (size_t i = 0; i < BN->getNumInputs(); i++) {
+        if (BN->getNthInput(i) == mean) {
```

Comment:
Could we directly modify the operands we believe are mean and var while asserting on them being equal instead of using a loop.
I.e.,
assert(BN->getNthInput(idxMean) ==  mean);
BN->setNthInput(idxMean, mean);

Reply:
Yeah, I could make that change if desired. Seems a little fragile as its based on the order of operands in NodeGen but the assert should catch any issues.

Reply:
Done.